### PR TITLE
Feature/improved exception handling

### PR DESCRIPTION
**Pull request by damniko**
The AstBuilder now gets source positions (line number and character position/index) from the CST generated by ANTLR.
Additionally, most exceptions have been modified to hopefully make them more modifiable (e.g., if we want to give a very specific error message in a very specific scenario), and exceptions are now "handled" in Main (by printing a formatted error message and stopping compilation).

Things to consider: Sometimes the compiler may be able to correct a mistake and continue the compilation. Since we want our compiler to be very explicit, the compilation stops completely when it encounters an error. This approach may seem a bit lazy, so we should defend it bravely when it inevitably is viciously attacked by Lone and her evil censor companion.

### CODE REVIEW

**Comment by damniko** [comment]
[Add line numbers, etc., to error messages in our compiler (transfer source positions from parser to our AST)](https://trello.com/c/0ZyW3vQU/119-add-line-numbers-etc-to-error-messages-in-our-compiler-transfer-source-positions-from-parser-to-our-ast)

**Review by Sarmisuper — APPROVED** [review]
lgtm
image

File: STEP/AST/PinTable.cs [hunk 1]
```diff
@@ -7,20 +7,22 @@ public class PinTable : IPinTable
     private readonly HashSet<int> _analogTable = new();
     private readonly HashSet<int> _digitalTable = new();
 
-    public void RegisterPin(TypeVal type, int pinVal)
+    public void RegisterPin(PinDclNode pinDcl)
```
Comment (WassawRoki) [inline STEP/AST/PinTable.cs:10]: explain this to me, PLZ

File: STEP/AST/SourcePosition.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
+﻿namespace STEP.AST;
+
+public class SourcePosition
+{
```
Comment (WassawRoki) [inline STEP/AST/SourcePosition.cs:4]: Så det bruges entlig, "bare" til at finde linje og position af en "fejl"

File: STEP/Program.cs [hunk 3]
```diff
@@ -29,17 +29,23 @@ private static void Main(string[] args)
 
         try
         {
+            Console.WriteLine("Performing syntactic analysis...");
             STEPParser.ProgramContext tree = parser.program(); // Parse the input starting at the "program" rule.
 
+            // TODO: error listener
+
```
Comment (WassawRoki) [inline STEP/Program.cs:60]: I like this: Det er ret fint det her med at printe hvad der sker, lidt som moderne loading :D

**Review by WassawRoki — APPROVED** [review]

File: STEP/Program.cs (hunk 3 above)
Comment (WassawRoki) [reply STEP/Program.cs:60]: hov skulle have været linjerne ovenfor

File: STEP/AST/PinTable.cs (hunk 1 above)
Comment (damniko) [reply STEP/AST/PinTable.cs:10]: Hvis vi tager imod en PinDclNode så får vi lidt mere info - f.eks. dens `SourcePosition`. Man kunne selvfølgelig også bare tilføje flere parametre, men i SymbolTable tager vi også mod nodes, så på den her måde er det consistent

File: STEP/AST/SourcePosition.cs (hunk 2 above)
Comment (damniko) [reply STEP/AST/SourcePosition.cs:4]: Yes! Så fejl-beskederne siger præcis hvor i kildekoden der er en fejl

File: STEP/Program.cs (hunk 3 above)
Comment (damniko) [reply STEP/Program.cs:60]: hehe

File: STEP/AST/SourcePosition.cs (hunk 2 above)
Comment (WassawRoki) [reply STEP/AST/SourcePosition.cs:4]: epic

**Comment by WassawRoki** [comment]
image

**Comment by WassawRoki** [comment]
image

**Pull request merged**